src/Uno.UI/UI/Xaml/Internal/EventManager.uno.cs

```diff
@@ -0,0 +1,140 @@
+﻿#nullable enable
+
+#if UNO_HAS_ENHANCED_LIFECYCLE
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.UI.Xaml;
+using Uno.UI.Helpers;
+using Windows.Foundation;
+
+namespace Uno.UI.Xaml.Core;
+
+// Code here is mostly matching the logic from WinUI (even if the code isn't really a direct port), but it's in CoreImports.cpp and LayoutManager.cpp
+// Since we don't yet have these, and this code is very related to events, we add it to this EventManager partial.
+internal sealed partial class EventManager
+{
+	private readonly ConditionalWeakTable<FrameworkElement, object?> _layoutUpdatedSubscribers = new();
+	private readonly Queue<SizeChangedQueueItem> _sizeChangedQueue = new(1024);
+
+	internal bool HasPendingSizeChangedEvents => _sizeChangedQueue is { Count: > 0 };
+
+	private readonly struct SizeChangedQueueItem : IEquatable<SizeChangedQueueItem>
+	{
+		public FrameworkElement Element { get; }
+		public Size OldSize { get; }
+
+		public SizeChangedQueueItem(FrameworkElement element, Size oldSize)
+		{
+			Element = element;
+			OldSize = oldSize;
+		}
+
+		public override bool Equals(object? obj)
+			=> obj is SizeChangedQueueItem item && Equals(item);
+
+		public bool Equals(SizeChangedQueueItem other)
+			=> ReferenceEquals(Element, other.Element) && OldSize == other.OldSize;
+
+		public override int GetHashCode()
+			=> HashCode.Combine(Element, OldSize);
+
+		public static bool operator ==(SizeChangedQueueItem left, SizeChangedQueueItem right)
+			=> left.Equals(right);
+
+		public static bool operator !=(SizeChangedQueueItem left, SizeChangedQueueItem right)
+			=> !(left == right);
+	}
+
+	public void AddLayoutUpdatedEventHandler(FrameworkElement element)
+	{
+		_layoutUpdatedSubscribers.Add(element, null);
+	}
+
+	public void RemoveLayoutUpdatedEventHandler(FrameworkElement element)
+	{
+		_layoutUpdatedSubscribers.Remove(element);
+	}
+
+	public void RaiseLayoutUpdated()
+	{
+		var clone = _layoutUpdatedSubscribers.ToArray();
+		for (int i = 0; i < clone.Length; i++)
+		{
+			clone[i].Key.OnLayoutUpdated();
+		}
```

Comment:
The `CollectionWeakTable` is [thread safe](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-8.0#thread-safety), including [the enumeration](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2.system-collections-generic-ienumerable-system-collections-generic-keyvaluepair-tkey-tvalue---getenumerator?view=net-8.0#remarks), you don't have to clone the collection (which should be avoided as this is a hot-path).

```suggestion
		foreach(var subscriber in _layoutUpdatedSubscribers)
		{
			subscriber.Key.OnLayoutUpdated();
		}
```

Reply:
This is not about thread-safety as it will happen on the UI thread anyways. It's more about re-entrancy.

Reply:
For reference: https://github.com/microsoft/microsoft-ui-xaml/blob/2a60e27c591846556fa9ec4d8f305afdf0f96dc1/dxaml/xcp/dxaml/lib/DXamlCore.cpp#L2179-L2215

Reply:
Could we use the array pool for that? This is likely going to be called a lot.

Reply:
Indeed I can see problem if you remove and then add back an handler for instance ... again some comments would help :)

(I still consider that we can avoid to clone the collection on something that can be considered as a hot-path, but if it replicates the code from winUI it's fine and we will change it only if we detect a performance hit)

Reply:
@jeromelaban I don't see a way to use array pool as ConditionalWeakTable doesn't provide a `Count` or `Length` property :/

We could go with counting it ourselves though, I think.

Reply:
Actually, adding elements while iterating isn't a problem. `ConditionalWeakTable` really does what we need and re-entrancy won't be an issue.

Reply:
Fixed now